File: backend/src/main/java/com/backend/api/question/service/QuestionService.java
package com.backend.api.question.service;

import com.backend.api.question.dto.request.QuestionAddRequest;
import com.backend.api.question.dto.request.QuestionUpdateRequest;
import com.backend.api.question.dto.response.*;
import com.backend.api.user.service.UserService;
import com.backend.domain.question.entity.Question;
import com.backend.domain.question.entity.QuestionCategoryType;
import com.backend.domain.question.repository.QuestionRepository;
import com.backend.domain.user.entity.Role;
import com.backend.domain.user.entity.User;
import com.backend.global.Rq.Rq;
import com.backend.global.exception.ErrorCode;
import com.backend.global.exception.ErrorException;
import jakarta.validation.Valid;
import lombok.RequiredArgsConstructor;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.PageRequest;
import org.springframework.data.domain.Pageable;
import org.springframework.data.domain.Sort;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;
import java.util.UUID;

@Service
@RequiredArgsConstructor
@Transactional(readOnly = true)
public class QuestionService {

    private final QuestionRepository questionRepository;
    private final UserService userService;
    private final Rq rq;

    //사용자 권한 검증
    private void validateUserAuthority(User user) {
        if (user == null) {
            throw new ErrorException(ErrorCode.UNAUTHORIZED_USER);
        }
        if (user.getRole() != Role.USER) {
            throw new ErrorException(ErrorCode.FORBIDDEN);
        }
    }

    //질문 존재 여부 검증
    public Question findByIdOrThrow(Long questionId) {
        return questionRepository.findById(questionId)
                .orElseThrow(() -> new ErrorException(ErrorCode.NOT_FOUND_QUESTION));
    }

    //질문 작성자 본인 여부 검증
    private void validateQuestionAuthor(Question question, User user) {
        if (!question.getAuthor().getId().equals(user.getId())) {
            throw new ErrorException(ErrorCode.FORBIDDEN);
        }
    }

    private void validateApprovedQuestion(Question question) {
        if (!question.getIsApproved()) {
            throw new ErrorException(ErrorCode.QUESTION_NOT_APPROVED);
        }
    }

    //사용자 질문 생성
    @Transactional
    public QuestionResponse addQuestion(@Valid QuestionAddRequest request, User user) {
        validateUserAuthority(user);
        Question question = createQuestion(request, user);
        Question saved = saveQuestion(question);
        return QuestionResponse.from(saved);
    }

    private Question createQuestion(QuestionAddRequest request, User user) {
        return Question.builder()
                .title(request.title())
                .content(request.content())
                .author(user)
                .categoryType(request.categoryType())
                .build();
    }


    private Question saveQuestion(Question question) {
        return questionRepository.save(question);
    }

    //사용자 질문 수정
    @Transactional
    public QuestionResponse updateQuestion(Long questionId, @Valid QuestionUpdateRequest request, User user) {
        validateUserAuthority(user);
        Question question = findByIdOrThrow(questionId);
        validateQuestionAuthor(question, user);
        updateQuestionContent(question, request);
        return QuestionResponse.from(question);
    }

    private void updateQuestionContent(Question question, QuestionUpdateRequest request) {
        question.updateUserQuestion(
                request.title(),
                request.content(),
                request.categoryType()
        );
    }

    //승인된 질문 전체 조회
    @Transactional(readOnly = true)
    public QuestionPageResponse<QuestionResponse> getApprovedQuestions(int page, QuestionCategoryType categoryType) {
        Page<Question> questionsPage;

        if (page < 1) page = 1;
        Pageable pageable = PageRequest.of(page - 1, 9, Sort.by("createDate").descending());

        if (categoryType == null) {
            questionsPage = questionRepository.findApprovedQuestionsExcludingCategory(QuestionCategoryType.PORTFOLIO, pageable);
        }
        else if (categoryType == QuestionCategoryType.PORTFOLIO) {
            // 포트폴리오만 조회
            questionsPage = questionRepository.findApprovedQuestionsByCategory(QuestionCategoryType.PORTFOLIO, pageable);
        }
        else {
            questionsPage = questionRepository.findByCategoryTypeAndIsApprovedTrue(categoryType, pageable);
        }

        if (questionsPage.isEmpty()) {
            throw new ErrorException(ErrorCode.NOT_FOUND_QUESTION);
        }

        List<QuestionResponse> questions = mapToResponseList(questionsPage);

        return QuestionPageResponse.from(questionsPage, questions);

    }

    private List<QuestionResponse> mapToResponseList(Page<Question> questionsPage) {
        return questionsPage.getContent()
                .stream()
                .map(QuestionResponse::from)
                .toList();
    }

    //승인된 질문 단건 조회
    @Transactional(readOnly = true)
    public QuestionResponse getApprovedQuestionById(Long questionId) {
        Question question = findByIdOrThrow(questionId);
        validateApprovedQuestion(question);
        return QuestionResponse.from(question);
    }

    //승인되지 않은 질문 단건 조회 -> 수정용
    @Transactional(readOnly = true)
    public QuestionResponse getNotApprovedQuestionById(Long userId, Long questionId, User user) {
        validateUserAuthority(user);

        if (!user.getId().equals(userId)) {
            throw new ErrorException(ErrorCode.QUESTION_INVALID_USER);
        }
        Question question = findByIdOrThrow(questionId);
        validateQuestionAuthor(question, user);

        if (question.getIsApproved()) {
            throw new ErrorException(ErrorCode.ALREADY_APPROVED_QUESTION);
        }

        return QuestionResponse.from(question);
    }

    @Transactional
    public void createListQuestion(List<Question> questions){
        questionRepository.saveAll(questions);
    }


    public AiQuestionReadAllResponse getByCategoryType(QuestionCategoryType questionCategoryType, User user) {
        return questionRepository.getQuestionByCategoryTypeAndUserId(questionCategoryType,user)
                .orElseThrow(() -> new ErrorException(ErrorCode.NOT_FOUND_QUESTION));
    }

    public PortfolioListReadResponse getByUserAndGroupId(User user, UUID groupId) {
        return questionRepository.getByUserAndGroupId(user, groupId)
                .orElseThrow(() -> new ErrorException(ErrorCode.NOT_FOUND_QUESTION));
    }

    @Transactional(readOnly = true)
    public int countByUser(User user) {
        validateUserAuthority(user);
        return questionRepository.countByAuthor(user);
    }

    @Transactional(readOnly = true)
    public QuestionPageResponse<QuestionResponse> findQuestionsByUserId(int page, Long userId) {
        userService.getUser(userId);
        User currentUser = rq.getUser();

        if (!currentUser.getId().equals(userId) && !currentUser.getRole().equals(Role.ADMIN)) {
            throw new ErrorException(ErrorCode.QUESTION_INVALID_USER);
        }

        if (page < 1) page = 1;
        Pageable pageable = PageRequest.of(page - 1, 15, Sort.by("createDate").descending());

        Page<Question> questionsPage = questionRepository.findByAuthorId(userId, pageable);

        List<QuestionResponse> questions = questionsPage
                .getContent()
                .stream()
                .map(QuestionResponse::from)
                .toList();

        return QuestionPageResponse.from(questionsPage, questions);
    }
}

File: backend/src/main/java/com/backend/api/question/service/QuestionService.kt
package com.backend.api.question.service

import com.backend.api.question.dto.request.QuestionAddRequest
import com.backend.api.question.dto.request.QuestionUpdateRequest
import com.backend.api.question.dto.response.AiQuestionReadAllResponse
import com.backend.api.question.dto.response.PortfolioListReadResponse
import com.backend.api.question.dto.response.QuestionPageResponse
import com.backend.api.question.dto.response.QuestionResponse
import com.backend.api.user.service.UserService
import com.backend.domain.question.entity.Question
import com.backend.domain.question.entity.QuestionCategoryType
import com.backend.domain.question.repository.QuestionRepository
import com.backend.domain.user.entity.Role
import com.backend.domain.user.entity.User
import com.backend.global.Rq.Rq
import com.backend.global.exception.ErrorCode
import com.backend.global.exception.ErrorException
import jakarta.validation.Valid
import org.springframework.data.domain.Page
import org.springframework.data.domain.PageRequest
import org.springframework.data.domain.Pageable
import org.springframework.data.domain.Sort
import org.springframework.data.repository.findByIdOrNull
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import java.util.*

@Service
@Transactional(readOnly = true)
class QuestionService(
    private val questionRepository: QuestionRepository,
    private val userService: UserService,
    private val rq: Rq
) {

    // 권한 검증
    private fun requireUser(user: User): User {
        if (user.role != Role.USER) throw ErrorException(ErrorCode.FORBIDDEN)
        return user
    }

    fun findByIdOrThrow(questionId: Long): Question =
        questionRepository.findByIdOrNull(questionId)
            ?: throw ErrorException(ErrorCode.NOT_FOUND_QUESTION)

    private fun validateQuestionAuthor(question: Question, user: User) {
        if (question.author.id != user.id) {
            throw ErrorException(ErrorCode.FORBIDDEN)
        }
    }

    private fun validateApprovedQuestion(question: Question) {
        if (!question.isApproved) {
            throw ErrorException(ErrorCode.QUESTION_NOT_APPROVED)
        }
    }

    @Transactional
    fun addQuestion(@Valid request: QuestionAddRequest, user: User): QuestionResponse {
        val u = requireUser(user)

        val question = Question(
            title = request.title,
            content = request.content,
            author = u,
            categoryType = request.categoryType
        )

        val saved = questionRepository.save(question)
        return QuestionResponse.from(saved)
    }

    @Transactional
    fun updateQuestion(questionId: Long, @Valid request: QuestionUpdateRequest, user: User): QuestionResponse {
        val u = requireUser(user)

        val question = findByIdOrThrow(questionId)
        validateQuestionAuthor(question, u)

        question.updateUserQuestion(
            request.title,
            request.content,
            request.categoryType
        )

        return QuestionResponse.from(question)
    }

    fun getApprovedQuestions(page: Int, categoryType: QuestionCategoryType?): QuestionPageResponse<QuestionResponse> {
        val pageNum = maxOf(page, 1)
        val pageable: Pageable = PageRequest.of(pageNum - 1, 9, Sort.by("createDate").descending())

        val questionsPage: Page<Question> = when (categoryType) {
            null ->
                questionRepository.findApprovedQuestionsExcludingCategory(
                    QuestionCategoryType.PORTFOLIO,
                    pageable
                )

            QuestionCategoryType.PORTFOLIO ->
                questionRepository.findApprovedQuestionsByCategory(
                    QuestionCategoryType.PORTFOLIO,
                    pageable
                )

            else ->
                questionRepository.findByCategoryTypeAndIsApprovedTrue(categoryType, pageable)
        }

        if (questionsPage.isEmpty) throw ErrorException(ErrorCode.NOT_FOUND_QUESTION)

        val responses = questionsPage.content.map { QuestionResponse.from(it) }
        return QuestionPageResponse.from(questionsPage, responses)
    }

    fun getApprovedQuestionById(questionId: Long): QuestionResponse {
        val question = findByIdOrThrow(questionId)
        validateApprovedQuestion(question)
        return QuestionResponse.from(question)
    }

    fun getNotApprovedQuestionById(userId: Long, questionId: Long, user: User): QuestionResponse {
        val u = requireUser(user)

        if (u.id != userId) throw ErrorException(ErrorCode.QUESTION_INVALID_USER)

        val question = findByIdOrThrow(questionId)
        validateQuestionAuthor(question, u)

        if (question.isApproved) {
            throw ErrorException(ErrorCode.ALREADY_APPROVED_QUESTION)
        }

        return QuestionResponse.from(question)
    }

    @Transactional
    fun createListQuestion(questions: List<Question>) {
        questionRepository.saveAll(questions)
    }

    fun getByCategoryType(categoryType: QuestionCategoryType, user: User): AiQuestionReadAllResponse =
        questionRepository.getQuestionByCategoryTypeAndUserId(categoryType, user)
            ?: throw ErrorException(ErrorCode.NOT_FOUND_QUESTION)

    fun getByUserAndGroupId(user: User, groupId: UUID): PortfolioListReadResponse =
        questionRepository.getByUserAndGroupId(user, groupId)
            ?: throw ErrorException(ErrorCode.NOT_FOUND_QUESTION)

    fun countByUser(user: User): Int {
        val u = requireUser(user)
        return questionRepository.countByAuthor(u)
    }

    fun findQuestionsByUserId(page: Int, userId: Long): QuestionPageResponse<QuestionResponse> {
        userService.getUser(userId) // 존재 확인

        val current = rq.getUser() // 이미 절대 non-null 로 설계되어 있음

        // 본인이거나 관리자만 가능
        if (current.id != userId && current.role != Role.ADMIN) {
            throw ErrorException(ErrorCode.QUESTION_INVALID_USER)
        }

        val pageNum = maxOf(page, 1)
        val pageable: Pageable = PageRequest.of(pageNum - 1, 15, Sort.by("createDate").descending())

        val questionsPage = questionRepository.findByAuthorId(userId, pageable)
        val responses = questionsPage.content.map { QuestionResponse.from(it) }

        return QuestionPageResponse.from(questionsPage, responses)
    }
}